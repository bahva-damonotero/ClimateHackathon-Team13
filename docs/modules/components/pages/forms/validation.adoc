= Validation

A component for listing the requirements of an input and checking them off as they are met.

== Usage Example

[tabs]
====
Preview::
+
--

++++
<iframe
  src="/uip/docs/examples-react/#/validation-example"
  width="100%"
  frameBorder="0"
  onload="resizeIframe(this)"
></iframe>
++++

--
JSX::
+
--

[source,jsx]
----
include::example$forms/ValidationExample.jsx[]
----

--
====

== Props
[grid="rows"]
|===
| *Name* | *Type* | *Default* | *Description*
| heading | `JSX.Element \| String` | | The heading content of the validation component.
include::../shared/shared.adoc[tag=id]
| requirements | `{ text: String, validator: String }[]` | | Array of requirements.
|===
