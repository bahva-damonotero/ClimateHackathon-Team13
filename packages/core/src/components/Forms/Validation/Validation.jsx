import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import validation from '@uswds/uswds/js/usa-validation';

export default function Validation({ children, heading, id, requirements }) {
  const containerRef = useRef();

  useEffect(() => {
    const containerEl = containerRef.current;
    if (!containerEl) return;

    // call USWDS's `on` function to initialize the component
    validation.on(containerEl);

    // on dismount, call USWDS's `off` function to cleanup the component
    return () => {
      validation.off(containerEl);
    };
  }, []);

  return (
    <div className="bah-uswds-validation" ref={containerRef}>
      <div className="usa-alert usa-alert--info usa-alert--validation">
        <div className="usa-alert__body">
          <h3 className="usa-alert__heading">{heading}</h3>

          <ul className="usa-checklist" id={id}>
            {requirements.map(({ text, validator }) => (
              <li
                className="usa-checklist__item"
                data-validator={validator}
                key={`${id}--${validator}`}
              >
                {text}
              </li>
            ))}
          </ul>
        </div>
      </div>

      {children}
    </div>
  );
}

Validation.propTypes = {
  children: PropTypes.node.isRequired,
  /** Heading content */
  heading: PropTypes.node.isRequired,
  /** ID of the validation */
  id: PropTypes.string.isRequired,
  /** Array of requirements */
  requirements: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      validator: PropTypes.string.isRequired
    })
  )
};
