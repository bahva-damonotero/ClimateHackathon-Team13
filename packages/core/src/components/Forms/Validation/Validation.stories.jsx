/* eslint-disable react/prop-types */
import React from 'react';

import Validation from '.';
import { Form, Input } from '..';

export default {
  title: 'Forms/Validation',
  component: Validation
};

const defaultArgs = {
  inputLabel: 'Code',
  inputId: 'validation-example',
  id: 'validation-id',
  heading: 'Enter a code'
};

const Template = ({ inputLabel, inputId, inputValidators, ...args }) => (
  <Form>
    <Validation {...args}>
      <Input id={inputId} validationId={args.id} validators={inputValidators}>
        {inputLabel}
      </Input>
    </Validation>
  </Form>
);

export const oneValidation = Template.bind({});
oneValidation.storyName = '1 Validation';
oneValidation.args = {
  ...defaultArgs,
  requirements: [
    {
      text: 'Use at least 8 characters',
      validator: 'length'
    }
  ],
  inputValidators: [
    {
      name: 'length',
      regex: '^.{8,}$'
    }
  ]
};

export const twoValidations = Template.bind({});
twoValidations.storyName = '2 Validations';
twoValidations.args = {
  ...defaultArgs,
  requirements: [
    {
      text: 'Use at least one uppercase character',
      validator: 'uppercase'
    },
    {
      text: 'Use at least one number',
      validator: 'numerical'
    }
  ],
  inputValidators: [
    {
      name: 'uppercase',
      regex: '[A-Z]'
    },
    {
      name: 'numerical',
      regex: '\\d'
    }
  ]
};
