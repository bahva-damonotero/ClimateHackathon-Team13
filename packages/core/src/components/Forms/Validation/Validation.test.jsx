import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import validation from '@uswds/uswds/js/usa-validation';

import Validation from '.';
import { Input } from '..';

describe('Validation', () => {
  const validationId = 'validation-example';

  const jsx = (
    <Validation
      heading="Validation test"
      id={validationId}
      requirements={[
        {
          text: 'Use at least 8 characters',
          validator: 'length-validation'
        }
      ]}
    >
      <Input
        id="validation-input"
        validationId={validationId}
        validators={[
          {
            name: 'length-validation',
            regex: '^.{8,}$'
          }
        ]}
      >
        Input
      </Input>
    </Validation>
  );

  it('should call `validation.on()` to initialize the component', () => {
    const onSpy = jest.spyOn(validation, 'on');

    render(jsx);

    expect(onSpy).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = renderer.create(jsx).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
