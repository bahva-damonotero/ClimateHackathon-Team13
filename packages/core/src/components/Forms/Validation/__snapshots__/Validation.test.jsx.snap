// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation should match snapshot 1`] = `
<div
  className="bah-uswds-validation"
>
  <div
    className="usa-alert usa-alert--info usa-alert--validation"
  >
    <div
      className="usa-alert__body"
    >
      <h3
        className="usa-alert__heading"
      >
        Validation test
      </h3>
      <ul
        className="usa-checklist"
        id="validation-example"
      >
        <li
          className="usa-checklist__item"
          data-validator="length-validation"
        >
          Use at least 8 characters
        </li>
      </ul>
    </div>
  </div>
  <label
    className="usa-label"
    htmlFor="validation-input"
  >
    Input
  </label>
  <input
    aria-describedby="validation-example"
    className="usa-input"
    data-validate-length-validation="^.{8,}$"
    data-validation-element="#validation-example"
    id="validation-input"
    name="validation-input"
    type="text"
  />
</div>
`;
